common/src/main/java/tachyon/worker/WorkerClient.java

```diff
@@ -236,8 +232,18 @@ private synchronized boolean connect() throws IOException {
         return false;
       }
       mConnected = true;
+
+      // only start the heartbeat thread if the connection is successful and if there is not
+      // another heartbeat thread running
+      if (mHeartbeat == null || mHeartbeat.isCancelled() || mHeartbeat.isDone()) {
```

Comment:
Only start a new HB task if there isn't one already running. Avoids that the HB threads creates a new self in cases of intermittent connection problems.